balancer/pickfirst/pickfirst.go

```diff
@@ -69,6 +77,238 @@ type pfConfig struct {
 	ShuffleAddressList bool `json:"shuffleAddressList"`
 }
 
+// subConnList stores provides functions to connect to a list of addresses using
+// the pick-first algorithm.
+type subConnList struct {
+	subConns []*scWrapper
+	// The index within the subConns list for the subConn being tried.
+	attemptingIndex int
+	b               *pickfirstBalancer
+	// The most recent failure during the initial connection attempt over the
+	// entire sunConns list.
+	lastFailure error
+	// Whether all the subConns have reported a transient failure once.
+	inTransientFailure bool
+	// Use a mutex to guard the list status as it can be changed concurrently by
+	// the idlePicker. Note that only calls to change the state to active are
+	// triggered from the picker and can happen concurrently.
+	mu     sync.RWMutex
+	status int
+}
+
+// scWrapper keeps track of the current state of the subConn.
+type scWrapper struct {
+	subConn  balancer.SubConn
+	conState connectivity.State
+	addr     resolver.Address
+}
+
+func newScWrapper(b *pickfirstBalancer, addr resolver.Address, listener func(state balancer.SubConnState)) (*scWrapper, error) {
+	scw := &scWrapper{
+		conState: connectivity.Idle,
+		addr:     addr,
+	}
+	sc, err := b.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{
+		StateListener: func(scs balancer.SubConnState) {
+			// Store the state and delegate.
+			scw.conState = scs.ConnectivityState
+			listener(scs)
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	scw.subConn = sc
+	return scw, nil
+}
+
+func newSubConnList(addrs []resolver.Address, b *pickfirstBalancer) *subConnList {
+	sl := &subConnList{
+		b:      b,
+		status: subConnListPending,
+	}
+
+	for _, addr := range addrs {
+		var scw *scWrapper
+		scw, err := newScWrapper(b, addr, func(state balancer.SubConnState) {
+			sl.stateListener(scw, state)
+		})
+		if err != nil {
+			if b.logger.V(2) {
+				b.logger.Infof("Ignoring failure, could not create a subConn for address %q due to error: %v", addr, err)
```

Comment:
Why is it safe to ignore the error here? In our existing code, if `NewSubConn` fails, we set the connectivity state to `TransientFailure`, return a picker that always returns an error (and includes the error returned from `NewSubConn`), and we also return `balancer.ErrBadResolverState`.

Reply:
This is what c-core seems to be doing. It throws an error only when we can't create a subConn for all the addresses: https://github.com/grpc/grpc/blob/a3a91a5a3cb6136d8a8d74e87b395f0ab450fd63/src/core/load_balancing/pick_first/pick_first.cc#L1017C5-L1026

I'll change it to fail fast.

Reply:
After thinking a little more, we use the next address when a connection to an address fails. By ignoring errors during creation of a subchannel for an address, we're treating the failure as a similar issue to a connection failure and using another address.
If we return an error early, we would miss the chance of using another address that we could create a subchannel for and use.